File: java/src/Bj12865.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Bj12865 {

    public static void solve() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] inputs = br.readLine().split(" ");
        int n = Integer.parseInt(inputs[0]);
        int k = Integer.parseInt(inputs[1]);

        int[][] dp = new int[n + 1][k + 1];

        for (int i = 1; i <= n; i++) {
            inputs = br.readLine().split(" ");
            int w = Integer.parseInt(inputs[0]);
            int v = Integer.parseInt(inputs[1]);

            for (int j = 1; j <= k; j++) {
                if (j >= w) {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - w] + v);
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
            }
        }

        System.out.println(dp[n][k]);
    }
}

File: kotlin/src/Bj12865.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.max

class Bj12865 {
    companion object{
        fun solve(){
            val br = BufferedReader(InputStreamReader(System.`in`))
            val (n, k) = br.readLine().split(" ").map { it.toInt() }

            val dp = Array(n + 1) { IntArray(k + 1) }

            for (i in 1..n) {
                val (w, v) = br.readLine().split(" ").map { it.toInt() }

                for (j in 1..k) {
                    if (j >= w) {
                        dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - w] + v)
                    } else {
                        dp[i][j] = dp[i - 1][j]
                    }
                }
            }

            println(dp[n][k])
        }
    }
}